internal/oci/container.go

```diff
@@ -347,18 +390,108 @@ func (c *Container) exitFilePath() string {
 func (c *Container) IsRunning() bool {
-	process, err := os.FindProcess(c.state.Pid)
+	_, err := c.pid()
 	if err != nil {
-		logrus.Infof("container %s not found: %v", c.id, err)
+		logrus.Errorf("checking if pid of %s is running failed: %v", c.id, err)
 		return false
 	}
 
-	if err := process.Signal(syscall.Signal(0)); err != nil {
-		logrus.Errorf("container %s not running %v", c.id, err)
-		return false
+	return true
+}
+
+// Pid is the external, locking function that returns the container's
+// pid, if it's running, and actually the container process
+func (c *Container) Pid() (int, error) {
+	c.opLock.Lock()
+	defer c.opLock.Unlock()
+	return c.pid()
+}
+
+// pid is the internal, non-locking version of pid. It checks
+// that we have a Pid defined in the state, that pid can be found
+// and it is the same process that was originally started by the runtime
+func (c *Container) pid() (int, error) {
+	if c.state == nil {
+		return -1, errors.New("state not initialized")
+	}
+	pid := c.state.InitialPid
+	if pid <= 0 {
+		return -1, errors.New("pid not initialized")
 	}
 
-	return true
+	// container has stopped (as pid is initialized but the runc state has overwritten it)
+	if c.state.Pid == 0 {
+		return 0, findprocess.ErrNotFound
+	}
+
+	if _, err := findprocess.FindProcess(pid); err != nil {
+		return -1, err
+	}
+
+	if err := c.verifyPid(pid); err != nil {
+		return -1, err
+	}
+	return pid, nil
+}
+
+// findAndReleasePid attempts to find the container's pid
+// and if it's found, release it. In between these steps,
+// it also verified the pid running is the same as was originally
+// started by the runtime
+func (c *Container) findAndReleasePid() (bool, error) {
+	pid := c.state.InitialPid
+	found := false
+
+	if err := c.verifyPid(pid); err != nil {
+		if !errors.Is(err, errContainerPidStopped) {
+			return found, err
+		}
+		return found, nil
+	}
+
+	process, err := findprocess.FindProcess(pid)
+	if err == findprocess.ErrNotFound {
+		return found, nil
+	}
+	if err != nil {
+		return found, err
+	}
+
+	found = true
+	if err := process.Release(); err != nil {
+		return found, err
+	}
+
+	return found, nil
+}
+
+// verifyPid checks that the start time for the process on the node is the same
+// as the start time we saved after creating the container.
+// This is the simplest way to verify we are operating on the container
+// process, and haven't run into pid wrap
+func (c *Container) verifyPid(pid int) error {
+	startTime, err := getProcessStartTime(pid)
+	if err != nil {
+		return err
+	}
+
+	if startTime != c.state.ProcessStartTime {
+		return errors.New("pid running but not the original container. pid wrap may have occurred")
+	}
+	return nil
+}
+
+// getProcessStartTime reads the kernel's /proc entry for stime for pid
+func getProcessStartTime(pid int) (string, error) {
+	psInfo, err := psgo.ProcessInfoByPids([]string{strconv.Itoa(pid)}, []string{"stime"})
```

Comment:
I would refrain from using `psgo` here. It reads five files in `/proc`, which is a bit too much for getting a single value.

If you only need a process start time, there is a trick you can use -- check any time (ctime, mtime, atime) of /proc/<PID> directory entry. This is a single call to `os.Stat()` and it will be way faster than parsing a few /proc files.